app/grants/utils.py

```diff
@@ -0,0 +1,548 @@
+from dashboard.models import Activity, Bounty, UserAction, Grant, Subscription, Contribution
+from eth_utils import to_checksum_address
+from gas.utils import conf_time_spread, eth_usd_conv_rate, gas_advisories, recommend_min_gas_price_to_confirm_in_time
+from hexbytes import HexBytes
+from ipfsapi.exceptions import CommunicationError
+from web3 import HTTPProvider, Web3
+from web3.exceptions import BadFunctionCallOutput
+
+logger = logging.getLogger(__name__)
+
+
+class GrantNotFoundException(Exception):
+    pass
+
+
+class UnsupportedNetworkException(Exception):
+    pass
+
+
+class IPFSCantConnectException(Exception):
+    pass
+
+
+class NoBountiesException(Exception):
+    pass
+
+
+def humanize_event_name(name):
+    """Humanize an event name.
+
+    Args:
+      name (str): The event name
+
+    Returns:
+        str: The humanized representation.
+
+    """
+    humanized_event_names = {
+        'new_bounty': 'New funded issue',
+        'start_work': 'Work started',
+        'stop_work': 'Work stopped',
+        'work_submitted': 'Work submitted',
+        'increased_bounty': 'Increased funds for issue',
+        'killed_bounty': 'Cancelled funded issue',
+        'worker_approved': 'Worker approved',
+        'worker_rejected': 'Worker rejected',
+        'work_done': 'Work done'
+    }
+
+    return humanized_event_names.get(name, name).upper()
+
+
+def create_user_action(user, action_type, request=None, metadata=None):
+    """Create a UserAction for the specified action type.
+
+    Args:
+        user (User): The User object.
+        action_type (str): The type of action to record.
+        request (Request): The request object. Defaults to: None.
+        metadata (dict): Any accompanying metadata to be added.
+            Defaults to: {}.
+
+    Returns:
+        bool: Whether or not the UserAction was created successfully.
+
+    """
+    from app.utils import handle_location_request
+    if action_type not in dict(UserAction.ACTION_TYPES).keys():
+        print('UserAction.create_action received an invalid action_type')
+        return False
+
+    if metadata is None:
+        metadata = {}
+
+    kwargs = {
+        'metadata': metadata,
+        'action': action_type,
+        'user': user
+    }
+
+    if request:
+        geolocation_data, ip_address = handle_location_request(request)
+
+        if geolocation_data:
+            kwargs['location_data'] = geolocation_data
+        if ip_address:
+            kwargs['ip_address'] = ip_address
+
+    if user and hasattr(user, 'profile'):
+        kwargs['profile'] = user.profile if user and user.profile else None
+
+    try:
+        UserAction.objects.create(**kwargs)
+        return True
+    except Exception as e:
+        logger.error(f'Failure in UserAction.create_action - ({e})')
+        return False
+
+
+def get_ipfs(host=None, port=settings.IPFS_API_PORT):
```

Comment:
No need to redefine all of these methods.  If you need any of these methods, import them from `app.utils` or `dashboard.utils`.  you can probably delete this whole file.